src/libclient/net/sessionlistingmodel.cpp

```diff
@@ -196,35 +229,60 @@ QVariant SessionListingModel::data(const QModelIndex &index, int role) const
 
 QVariant SessionListingModel::headerData(int section, Qt::Orientation orientation, int role) const
 {
-	if(role != Qt::DisplayRole || orientation != Qt::Horizontal)
+	if(orientation != Qt::Horizontal)
 		return QVariant();
 
-	switch(section) {
-	case ColumnVersion: return QVariant();
-	case ColumnStatus: return QVariant();
-	case ColumnTitle: return tr("Title");
-	case ColumnServer: return tr("Server");
-	case ColumnUsers: return tr("Users");
-	case ColumnOwner: return tr("Owner");
-	case ColumnUptime: return tr("Age");
+	if (role == Qt::AccessibleTextRole && section == Version) {
+		return tr("Compatibility");
+	} else if (role != Qt::DisplayRole) {
+		return QVariant();
 	}
-	static_assert (ColumnCount == 7, "Update headerData()");
 
+	switch(Column(section)) {
+	case Title: return tr("Title");
+	case Server: return tr("Server");
+	case UserCount: return tr("Users");
+	case Owner: return tr("Owner");
+	case Uptime: return tr("Age");
+	case Version:
+	case ColumnCount: {}
+	}
 	return QVariant();
 }
 
 Qt::ItemFlags SessionListingModel::flags(const QModelIndex &index) const
 {
-	if(index.internalId()>0) {
-		const Listing &l = m_listings.at(int(index.internalId() - 1));
-		if(!l.message.isEmpty())
-			return Qt::NoItemFlags;
-
-		const Session &s = l.sessions.at(index.row());
-		if(!s.protocol.isCompatible())
-			return Qt::NoItemFlags;
+	if (isRootItem(index)) {
+		return Qt::ItemIsEnabled;
+	}
+
+	const auto &listing = m_listings.at(listingIndex(index));
+	if(listing.offline())
+		return Qt::ItemIsEnabled | Qt::ItemNeverHasChildren;
+
+	const auto &session = listing.sessions.at(index.row());
```

Comment:
Getting an index out of bounds right here. Adding a check leads to a busted server list where all items are disabled and toggling any of the filters takes an eternity to apply, so it seems like it's just going way off the deep end somewhere.

Reproduction setup: start with an empty set of list servers, Session → Join, type `pub.drawpile.net` into the input, click Add Server, accept the terms and conditions.

The error reproduces every time the session list is loaded from the list server.

Reply:
Hrm, I did a clean build now and it doesn't crash anymore, so maybe the moc just got messed up or something. So actually, the listings now don't crash.

Reply:
This is fixed now and was due to an off-by-one error in the model caused by Qt’s use of inclusive ranges.